energymeters/integrationpluginenergymeters.cpp

```diff
@@ -0,0 +1,394 @@
+{
+    qCDebug(dcEnergyMeters) << "Post setup thing" << thing->name();
+    if (m_connectionStateTypeIds.contains(thing->thingClassId())) {
+        thing->setStateValue(m_connectionStateTypeIds.value(thing->thingClassId()), true);
+
+        if (m_energyMeters.contains(thing)) {
+            startUpdateCycle(m_energyMeters.value(thing));
+        }
+    }
+
+    if (!m_reconnectTimer) {
+        m_reconnectTimer = hardwareManager()->pluginTimerManager()->registerTimer(5);
+        connect(m_reconnectTimer, &PluginTimer::timeout, this, [this] {
+            foreach (Thing *thing, myThings()) {
+                if (m_connectionStateTypeIds.contains(thing->thingClassId())) {
+                    if (!thing->stateValue(m_connectionStateTypeIds.value(thing->thingClassId())).toBool()) {
+                        EnergyMeter *meter = m_energyMeters.value(thing);
+                        if (!meter)
+                            continue;
+
+                        startUpdateCycle(meter);
+                    }
+                }
+            }
+        });
+    }
+}
+
+void IntegrationPluginEnergyMeters::thingRemoved(Thing *thing)
+{
+    qCDebug(dcEnergyMeters()) << "Thing removed" << thing->name();
+
+    if (m_energyMeters.contains(thing)) {
+        EnergyMeter *meter = m_energyMeters.take(thing);
+        m_updateCycleInProgress.remove(meter);
+        meter->deleteLater();
+    }
+
+    if (myThings().isEmpty() && m_reconnectTimer) {
+        qCDebug(dcEnergyMeters()) << "Stopping reconnect timer";
+        hardwareManager()->pluginTimerManager()->unregisterTimer(m_reconnectTimer);
+        m_reconnectTimer = nullptr;
+    }
+}
+
+void IntegrationPluginEnergyMeters::startUpdateCycle(EnergyMeter *meter)
+{
+    if (m_updateCycleInProgress.contains(meter)) {
+        if (m_updateCycleInProgress.value(meter)) {
+            return;
+        }
+    }
+    m_updateCycleInProgress.insert(meter, true);
+    meter->getVoltageL1();
+}
+
+void IntegrationPluginEnergyMeters::updateCycleFinished(EnergyMeter *meter)
+{
+    m_updateCycleInProgress.insert(meter, false);
+
+    int updateInterval = configValue(energyMetersPluginUpdateIntervalParamTypeId).toInt();
+    QTimer::singleShot(updateInterval, meter, [this, meter] {
+        startUpdateCycle(meter); // restart update cycle
+    });
+}
+
+void IntegrationPluginEnergyMeters::onConnectionStateChanged(bool status)
+{
+    EnergyMeter *meter = static_cast<EnergyMeter *>(sender());
+    Thing *thing = m_energyMeters.key(meter);
+    if (!thing)
+        return;
+
+    if (!status) {
+        updateCycleFinished(meter);
+    }
+    thing->setStateValue(m_connectionStateTypeIds.value(thing->thingClassId()), status);
+}
+
+void IntegrationPluginEnergyMeters::onVoltageL1Received(double voltage)
+{
+    EnergyMeter *meter = static_cast<EnergyMeter *>(sender());
+    Thing *thing = m_energyMeters.key(meter);
+    if (!thing)
+        return;
+
+    meter->getVoltageL2();
```

Comment:
Not sure if this works? `meter->getVoltageL2()` will start an async call, and you set the state right after the call? Isn't this creating an infinit loop? on voltage received -> get voltage -> on voltage received...